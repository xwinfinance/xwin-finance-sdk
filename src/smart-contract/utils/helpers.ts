import {JsonRpcProvider, Wallet, ethers} from 'ethers';
import {IGasFee} from './interfaces';

export const getJsonRpcProvider = async ({
  rpcNodeUrl,
  privateKey,
}: {
  rpcNodeUrl: string;
  privateKey?: string;
}): Promise<{
  provider: JsonRpcProvider;
  chainId: Number;
  signer?: Wallet;
}> => {
  const provider = new JsonRpcProvider(rpcNodeUrl);
  const chainId = Number((await provider.getNetwork()).chainId);

  if (!privateKey) {
    return {
      provider,
      chainId,
    };
  }

  return {
    provider,
    chainId,
    signer: new Wallet(privateKey, provider),
  };
};

export const convertToWei = (val: number): string => {
  return ethers.parseUnits(val.toString(), 'gwei').toString();
};

export const getFastGasFee = async (chainId: Number): Promise<IGasFee> => {
  // BNB Smart Chain Mainnet
  if (chainId === 56) {
    return {
      gasPrice: convertToWei(3),
    };
  }

  // Polygon Mainnet
  if (chainId === 137) {
    const gasData = await (await fetch('https://gasstation-mainnet.matic.network/v2')).json();
    return {
      maxFeePerGas: convertToWei(gasData.fast.maxFee.toFixed(6)),
      maxPriorityFeePerGas: convertToWei(gasData.fast.maxPriorityFee.toFixed(6)),
    };
  }

  // Mumbai - Testnet
  if (chainId === 80001) {
    const gasData = await (await fetch('https://gasstation-mumbai.matic.today/v2')).json();
    return {
      maxFeePerGas: convertToWei(gasData.fast.maxFee.toFixed(6)),
      maxPriorityFeePerGas: convertToWei(gasData.fast.maxPriorityFee.toFixed(6)),
    };
  }

  // BNB Smart Chain Testnet
  if (chainId === 97) {
    return {
      gasPrice: convertToWei(10),
    };
  }

  // GoChain Testnet
  if (chainId === 31337) {
    return {
      gasPrice: convertToWei(0.1),
    };
  }

  // Arbitrum One
  if (chainId === 42161) {
    return {
      gasPrice: convertToWei(0.1),
    };
  }

  return {
    gasPrice: convertToWei(3),
  };
};
