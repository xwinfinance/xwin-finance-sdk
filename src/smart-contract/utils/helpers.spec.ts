import {convertSlippage, convertToWei, getFastGasFee, getJsonRpcProvider} from './helpers';

describe('helpers', () => {
  describe('getJsonRpcProvider', () => {
    it('resolves as expected - Binance', async () => {
      // Pre

      // Action
      const output = await getJsonRpcProvider({
        rpcNodeUrl: 'https://bsc-dataseed1.binance.org',
      });
      // Assert

      expect(output).toMatchSnapshot();
    });

    it('resolves as expected - polygon', async () => {
      // Pre

      // Action
      const output = await getJsonRpcProvider({
        rpcNodeUrl: 'https://polygon-rpc.com',
      });
      // Assert

      expect(output).toMatchSnapshot();
    });

    it('resolves as expected - arbitrum', async () => {
      // Pre

      // Action
      const output = await getJsonRpcProvider({
        rpcNodeUrl: 'https://arbitrum-one.publicnode.com',
      });
      // Assert

      expect(output).toMatchSnapshot();
    });
  });

  describe('convertToWei', () => {
    it('resolves as expected', () => {
      // Pre
      const input = 3;

      // action
      const output = convertToWei(input);

      // Assert
      expect(output).toBe('3000000000');
    });
  });

  describe('getFastGasFee', () => {
    it.each([
      [{chainId: 56}],
      [{chainId: 97}],
      [{chainId: 31337}],
      [{chainId: 42161}],
      [{chainId: 10}],
    ])('resolves as expected by %s', async (val) => {
      // Pre
      // action
      const output = await getFastGasFee(val.chainId);

      // Assert
      expect(output).toMatchSnapshot();
    });
  });

  describe('convertSlippage', () => {
    it('resolves as expected', () => {
      // Pre
      // Action
      const output = convertSlippage(0.2);
      // Assert
      expect(output).toMatchSnapshot();
    });

    it('throw error when value > 1', () => {
      // Pre
      try {
        // Action
        convertSlippage(1.2);
      } catch (e) {
        // Assert
        expect(e).toBeInstanceOf(Error);
        expect(e.message).toBe('Slippage value must be not greater than 1');
      }
    });
  });
});
