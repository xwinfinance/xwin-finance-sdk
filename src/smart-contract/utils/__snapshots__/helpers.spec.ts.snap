// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFastGasFee resolves as expected by { chainId: 10 } 1`] = `
{
  "gasPrice": "3000000000",
}
`;

exports[`getFastGasFee resolves as expected by { chainId: 56 } 1`] = `
{
  "gasPrice": "3000000000",
}
`;

exports[`getFastGasFee resolves as expected by { chainId: 97 } 1`] = `
{
  "gasPrice": "10000000000",
}
`;

exports[`getFastGasFee resolves as expected by { chainId: 31337 } 1`] = `
{
  "gasPrice": "100000000",
}
`;

exports[`getFastGasFee resolves as expected by { chainId: 42161 } 1`] = `
{
  "gasPrice": "100000000",
}
`;

exports[`getJsonRpcProvider resolves as expected - Binance 1`] = `
{
  "chainId": 56,
  "provider": JsonRpcProvider {},
}
`;

exports[`getJsonRpcProvider resolves as expected - arbitrum 1`] = `
{
  "chainId": 42161,
  "provider": JsonRpcProvider {},
}
`;

exports[`getJsonRpcProvider resolves as expected - polygon 1`] = `
{
  "chainId": 137,
  "provider": JsonRpcProvider {},
}
`;
